doc/developer/design/20240411_graceful_reconfiguration_of_managed_clusters_v1.md

```diff
@@ -0,0 +1,329 @@
+replicas and move the reconfig-replicas to be active. Finally remove the
+Reconfigure schedule from the cluster.
+
+
+#### Cluster Sequencer
+The following roughly defines the new logic for the cluster sequencer when performing 
+an alter cluster on a managed cluster.
+````
+NOTE:
+This assumes we're storing prior config in the reconfig schedule, if we're
+not we can probably just compare against the current config in which case we
+wouldn't allow any cancelation.
+````
+
+In sequence_alter_cluster_managed_to_managed:
+ Is there a reconfiguration going on.. 
+   - no
+      - if delay = false
+        - use current logic
+      - if dealy = true
+        - create new replicas (but do not drop the old ones).
+        - set the cluster reconfiugration scheulde to specified duraiton (defaults to 0)
+   - yes
+     - if delay = true
+       - if new config = prior config (ex: new size = size stored in `ClusterScheudle::Reconfigure.prior_config`)
+         -  The user is effectively cancelling an in-progress
+            reconfiguration, drop the reconf replicas. (And possibly adjust
+            the number of stable replicas if the replication factor has changed.)
+            reset the cluster schedule.
+       - if config = current_config (ex: new size = current catalog size)
+         - adjust the number of reconf replicas if the replication factor changed and
+           adjust the reconfigure schedule deadline if it has changed.
+          // POTENTIAL ISSUE: Using reconfiguration deadline seems like it
+          //  may run into some issues. If the same command is run twice do
+          //  we just bump up the deadline? It does have the nice behavior
+          //  where you could set the timeout to 0 and force an early
+          //  reoncifigure_finalization.
+       - if new config != prior or current
+         -  drop the reconf replicas, and create new reconf replicas.
+            update the reconfiguration deadline as needed
+     - if delay = false
+       - follow the old logic of dropping and re-creating, drop all replicas,
+         including reconfig replcias, reset the cluster schedule.
+```
+CONCERN:
+Allowing delay=false and a timeout=0 provides two similar mechanisms of
+using the old behavior. I think this is what we want as it ensures the old
+replicas are destroyed before the new ones are created which has billining
+implications. If we were to always rely on the reconfiguration_finalization
+we may charge for both sets of replicas for some additional time if the
+coordinator has a stall. If we're not concerned about this, always using
+reconfiguration_finalization may be better as it reduces the code paths.
+```
+
+## Open Questions
+
+### How do we handle multiple schedules?
+If this feature is introducing multiple schedules we'll need some rules governing the interaction between schedules.
+There's world where each schedule needs to consider it's interaction with all other schedules and that seems kind of
+scary.
```

Comment:
I think there could be a division of schedules into 2 categories:
1. Schedules that turn clusters On/Off.
2. Schedules that affect the set of replicas of a cluster, _if the cluster is On_.

The interaction among all the schedules in the 1. category are already handled by the current code. (We wait for all schedules to have their say about a cluster before touching a cluster, and the cluster will be On if at least one schedule says that it should be On.)

The interaction between 1. and 2. would be:
- If 1. decided that the cluster is Off, then 2. has nothing to do, because there are no replicas.
- If 1. decided that the cluster is On, then 2. can decide on the exact set of replicas.

This leaves open the interactions among different schedules in category 2. What would be another example of a schedule in 2.? Maybe auto-scaling? I guess that would need to have some custom-defined interaction with graceful reconf timeout schedule.